python/tests/test_ownership_dataset.py

```diff
@@ -0,0 +1,275 @@
+    da.coords['x']['x', 0] = -1
+    da.attrs['a']['x', 0] = -100
+    da.masks['m']['x', 0] = False
+    c['x', 1] = -2
+    a['x', 1] = -200
+    m['x', 1] = True
+    da.unit = 'kg'
+    da.coords['x'].unit = 'J'
+    assert sc.identical(
+        da,
+        sc.DataArray(
+            sc.array(dims=['x'], values=[-10, -20], unit='kg'),
+            coords={'x': sc.array(dims=['x'], values=[-1, -2], unit='J')},
+            attrs={'a': sc.array(dims=['x'], values=[-100, -200])},
+            masks={'m': sc.array(dims=['x'], values=[False, True])}))
+    assert sc.identical(v, sc.array(dims=['x'], values=[-10, -20], unit='kg'))
+    assert sc.identical(c, sc.array(dims=['x'], values=[-1, -2], unit='J'))
+    assert sc.identical(a, sc.array(dims=['x'], values=[-100, -200]))
+    assert sc.identical(m, sc.array(dims=['x'], values=[False, True]))
+
+    # TODO intentional?
+    # Assignments overwrite data but not metadata.
+    da.data = sc.array(dims=['x'], values=[11, 22], unit='m')
+    da.coords['x'] = sc.array(dims=['x'], values=[3, 4], unit='s')
+    da.attrs['a'] = sc.array(dims=['x'], values=[300, 400])
+    da.masks['m'] = sc.array(dims=['x'], values=[True, True])
+    assert sc.identical(
+        da,
+        sc.DataArray(
+            sc.array(dims=['x'], values=[11, 22], unit='m'),
+            coords={'x': sc.array(dims=['x'], values=[3, 4], unit='s')},
+            attrs={'a': sc.array(dims=['x'], values=[300, 400])},
+            masks={'m': sc.array(dims=['x'], values=[True, True])}))
+    assert sc.identical(v, sc.array(dims=['x'], values=[11, 22], unit='m'))
+    assert sc.identical(c, sc.array(dims=['x'], values=[-1, -2], unit='J'))
+    assert sc.identical(a, sc.array(dims=['x'], values=[-100, -200]))
+    assert sc.identical(m, sc.array(dims=['x'], values=[False, True]))
+
+
+def test_own_darr_get():
+    # Data and metadata are shared.
+    da = make_data_array()[0]
+    v = da.data
+    c = da.coords['x']
+    a = da.attrs['a']
+    m = da.masks['m']
+    da['x', 0] = -10
+    da.data['x', 1] = -20
+    da.coords['x']['x', 0] = -1
+    da.attrs['a']['x', 0] = -100
+    da.masks['m']['x', 0] = False
+    c['x', 1] = -2
+    a['x', 1] = -200
+    m['x', 1] = True
+    da.unit = 'kg'
+    da.coords['x'].unit = 'J'
+    assert sc.identical(
+        da,
+        sc.DataArray(
+            sc.array(dims=['x'], values=[-10, -20], unit='kg'),
+            coords={'x': sc.array(dims=['x'], values=[-1, -2], unit='J')},
+            attrs={'a': sc.array(dims=['x'], values=[-100, -200])},
+            masks={'m': sc.array(dims=['x'], values=[False, True])}))
+    assert sc.identical(v, sc.array(dims=['x'], values=[-10, -20], unit='kg'))
+    assert sc.identical(c, sc.array(dims=['x'], values=[-1, -2], unit='J'))
+    assert sc.identical(a, sc.array(dims=['x'], values=[-100, -200]))
+    assert sc.identical(m, sc.array(dims=['x'], values=[False, True]))
+
+    # TODO intentional?
+    # Assignments overwrite data but not coords.
```

Comment:
Intentional as above: When we use `__getitem__` we share ownership of the meta data. `__setitem__` does not assign, otherwise we would prevent, e.g., replacing a coord with one of a different dtype.